[WGSL] Port WGSLUnitTests to TestWebKitAPI

djg opened the pull request:
#### 5bda1cab2367440c3b9bb734234b75d876d3e6c1
<pre>
[WGSL] Port WGSLUnitTests to TestWebKitAPI
<a href="https://bugs.webkit.org/show_bug.cgi?id=245061">https://bugs.webkit.org/show_bug.cgi?id=245061</a>
rdar://problem/99808942

Reviewed by Elliott Williams.

So we can run the tests on EWS, port the XCTest-based WGSLUnitTests to
gtest-based standalone executable TestWGSL. This was a straight-forward change
of ObjC interfaces to C++ classes and renaming of testing macros.

* Tools/Scripts/run-api-tests:
* Tools/Scripts/webkitpy/port/darwin.py:
(DarwinPort):
* Tools/TestWebKitAPI/Configurations/TestWGSL.xcconfig: Added.
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/xcshareddata/xcschemes/TestWGSL.xcscheme: Added.
* Tools/TestWebKitAPI/Tests/WGSL/LexerTests.cpp: Added.
(TestWebKitAPI::checkSingleToken):
(TestWebKitAPI::checkSingleLiteral):
(TestWebKitAPI::checkNextTokenIs):
(TestWebKitAPI::checkNextTokenIsIdentifier):
(TestWebKitAPI::checkNextTokenIsLiteral):
(TestWebKitAPI::checkNextTokensAreBuiltinAttr):
(TestWebKitAPI::TEST):
* Tools/TestWebKitAPI/Tests/WGSL/ParserTests.cpp: Added.
(TestWebKitAPI::TEST):
* Tools/TestWebKitAPI/mac/mainMac.mm:
(main):

Canonical link: <a href="https://commits.webkit.org/254654@main">https://commits.webkit.org/254654@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b3d5e4c28b74eaf8a6217117fdeefd8cd073b402

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/89655 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/34206 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/20371 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/98947 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/155634 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/93663 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/32700 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/28174 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/82017 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/93355 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/95302 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/68/builds/25982 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/76512 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/25931 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/5/builds/93241 "Passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/80859 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/80740 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/68925 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/30455 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/14822 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/30205 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/15754 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3268 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/33653 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/38700 "Found 1 new test failure: imported/w3c/web-platform-tests/html/semantics/embedded-content/the-iframe-element/sandbox-inherit-to-blank-document-unsandboxed.html (failure)") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/32363 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/34845 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

emw-apple commented on Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
```diff
@@ -5745,7 +5759,7 @@
 				"$(SRCROOT)/Sources.txt",
 				"$(SRCROOT)/SourcesCocoa.txt",
-				"$(BUILT_PRODUCTS_DIR)/usr/local/include/wtf/Scripts/generate-unified-source-bundles.rb",
+				"$(SRCROOT)../../Source/wtf/Scripts/generate-unified-source-bundles.rb",
```
I think this should be:
```suggestion
				"$(SRCROOT)../../Source/WTF/Scripts/generate-unified-source-bundles.rb",
```

Also, what's the reason for changing this? It does seem valid these days (we would never build TestWebKitAPI without a full checkout) but plumbing scripts through WTF's include directory is more typical.

djg replied:
I've removed this. I think it was from me trying to work around an error. I didn't realise that it was necessary to build WebKit first because TestWebKitAPI doesn't have correct dependencies to build everything from a clean slate.

emw-apple commented on Tools/TestWebKitAPI/Configurations/Base.xcconfig:
```diff
@@ -37,7 +37,7 @@ CLANG_WARN_CXX0X_EXTENSIONS = NO;
 ALTERNATE_HEADER_SEARCH_PATHS_iosmac = $(BUILT_PRODUCTS_DIR)$(WK_ALTERNATE_FRAMEWORKS_DIR)/usr/local/include
 
-PROJECT_HEADER_SEARCH_PATHS = $(SRCROOT)/../../Source/WebKit/Platform/cocoa $(PROJECT_HEADER_SEARCH_PATHS_$(WK_COCOA_TOUCH));
+PROJECT_HEADER_SEARCH_PATHS = $(SRCROOT)/../../Source/WebKit/Platform/cocoa $(SRCROOT)/../../Source/WebGPU $(PROJECT_HEADER_SEARCH_PATHS_$(WK_COCOA_TOUCH));
```
Are there specific imports that this fixes? Asking because this PR changes many WGSL quote imports to library imports, so I want to make sure this is still needed.

djg replied:
I was having troubles getting another xcodeproj to access the WGSL headers because WebGPU.xcodeproj uses header maps. I eventually promoted the internal headers to private headers so they are installed in $(BUILT_PRODUCTS_DIR) and TestWebKitAPI can use flat `#include <WGSL/header.h>`.

Please let me know if there is a better way to do this!

emw-apple reviewed (commented):
One high-level question I have is whether we should consider creating a separate test executable for WGSL, like @kkinnunen-apple did with TestIPC in https://github.com/WebKit/WebKit/pull/3360. Maybe it doesn't make sense for only two test classes?

That aside, this all seems reasonable.

kkinnunen-apple commented on Tools/Scripts/run-api-tests:
```diff
@@ -117,6 +117,8 @@ def parse_args(args):
         optparse.make_option('--webkit-legacy-only', action='store_const', const='TestWebKitLegacy', dest='api_binary',
                              help='Only check and run TestWebKitLegacy.exe (Windows only)'),
+        optparse.make_option('--wgsl-only', action='store_const', const='TestWGSL', dest='api_binary',
+                             help='Only check and run TestWGSL.exe (Darwin ports only)'),
```
Probably not TestWGSL.exe, just TestWGSL

kkinnunen-apple commented on Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
```diff
@@ -1335,6 +1387,16 @@
+			buildActionMask = 2147483647;
+			dstPath = "";
+			dstSubfolderSpec = 16;
+			files = (
```
I think you need to add WebKit.framework to product build dependencies, as is added below.

The TestWebKitAPI/config.h uses headers generated by WebKit.framework build.

djg replied:
OK. I wasn't using anything from `WebKit.framework` so I removed it.

kkinnunen-apple reviewed (commented):
Looks fine by me, I'll let Elliott make the judgement

emw-apple commented on Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
```diff
@@ -5444,6 +5558,26 @@
+			buildRules = (
+			);
+			dependencies = (
+				3A5DDAEC28D156BB004DA950 /* PBXTargetDependency */,
```
Is this target dependency on WGSL necessary? Having it means that any attempt to build TestWebKitAPI will also build WGSL, including `make` builds and production builds, which could be a problem. Can we remove it?

Because TestWGSL links against libwgsl.a (on L3496), there is already an implicit dependency. So any scheme with "Find Implicit Dependencies" turned on will build WGSL and TestWGSL in the correct order.

emw-apple commented on Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
```diff
@@ -6567,6 +6762,22 @@
+			isa = XCBuildConfiguration;
+			baseConfigurationReference = 3A5DDADB28D15328004DA950 /* TestWGSL.xcconfig */;
+			buildSettings = {
+				PRODUCT_NAME = "$(TARGET_NAME)";
```
This is added by default in Xcode, but not needed, since we put everything in xcconfigs. I usually clean these up by going into the target's build settings panel, selecting "Customized", selecting all and pressing Delete.

emw-apple commented on Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
```diff
@@ -6567,6 +6762,22 @@
+			isa = XCBuildConfiguration;
+			baseConfigurationReference = 3A5DDADB28D15328004DA950 /* TestWGSL.xcconfig */;
+			buildSettings = {
+				PRODUCT_NAME = "$(TARGET_NAME)";
```
Ditto above.

emw-apple commented on Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
```diff
@@ -3489,6 +3577,7 @@
 		08FB7794FE84155DC02AAC07 /* TestWebKitAPI */ = {
 			isa = PBXGroup;
 			children = (
+				3A5DDADC28D156AF004DA950 /* WebGPU.xcodeproj */,
```
Assuming you're able to remove the explicit target dependency on WGSL, please also delete the project reference Xcode added. That will remove quite a bit of noise from the pbxproj diff :)

emw-apple commented on Tools/TestWebKitAPI/Configurations/TestWGSL.xcconfig:
```diff
@@ -0,0 +1,44 @@
+GCC_ENABLE_OBJC_EXCEPTIONS = YES;
+
+GCC_PREPROCESSOR_DEFINITIONS = $(inherited) BUILDING_TEST_WGSL GTEST_API_=
+
```
In https://github.com/WebKit/WebKit/commit/a85b766e565b6a913b3fed1075d645934cd47cad, TestIPC was broken and I had to add 

```
// Workaround for "WebKit is not available..." error (rdar://91668054)
DISABLE_SDK_METADATA_PARSING[sdk=appletv*] = YES;
DISABLE_SDK_METADATA_PARSING[sdk=watch*] = YES;
```
to fix the tvOS and watchOS builds. Maybe we need the same here? I think the underlying Xcode issue was fixed in Xcode 14, so we might be safe now, but just calling it out in case you see any failures.

djg committed:
[WGSL] Port WGSLUnitTests to TestWebKitAPI
https://bugs.webkit.org/show_bug.cgi?id=245061
rdar://problem/99808942

Reviewed by Elliott Williams.

So we can run the tests on EWS, port the XCTest-based WGSLUnitTests to
gtest-based standalone executable TestWGSL. This was a straight-forward change
of ObjC interfaces to C++ classes and renaming of testing macros.

* Tools/Scripts/run-api-tests:
* Tools/Scripts/webkitpy/port/darwin.py:
(DarwinPort):
* Tools/TestWebKitAPI/Configurations/TestWGSL.xcconfig: Added.
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/xcshareddata/xcschemes/TestWGSL.xcscheme: Added.
* Tools/TestWebKitAPI/Tests/WGSL/LexerTests.cpp: Added.
(TestWebKitAPI::checkSingleToken):
(TestWebKitAPI::checkSingleLiteral):
(TestWebKitAPI::checkNextTokenIs):
(TestWebKitAPI::checkNextTokenIsIdentifier):
(TestWebKitAPI::checkNextTokenIsLiteral):
(TestWebKitAPI::checkNextTokensAreBuiltinAttr):
(TestWebKitAPI::TEST):
* Tools/TestWebKitAPI/Tests/WGSL/ParserTests.cpp: Added.
(TestWebKitAPI::TEST):
* Tools/TestWebKitAPI/mac/mainMac.mm:
(main):

Canonical link: https://commits.webkit.org/254654@main
